controllers/authorino_controller.go

```diff
@@ -0,0 +1,598 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/go-logr/logr"
+	authorinooperatorv1beta1 "github.com/kuadrant/authorino-operator/api/v1beta1"
+)
+
+// AuthorinoReconciler reconciles a Authorino object
+type AuthorinoReconciler struct {
+	client.Client
+	OperatorNamespace string
+	Log               logr.Logger
+	Scheme            *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=authorino-operator.kuadrant.3scale.net,resources=authorinoes,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=authorino-operator.kuadrant.3scale.net,resources=authorinoes/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=authorino-operator.kuadrant.3scale.net,resources=authorinoes/finalizers,verbs=update
+
+// +kubebuilder:rbac:groups="apps",resources=deployments,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Authorino object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.2/pkg/reconcile
+func (r *AuthorinoReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("authorinoReconciler", req.NamespacedName)
+
+	//get operator namespace
+	r.OperatorNamespace = authorinooperatorv1beta1.AuthorinoOperatorNamespace
+	ns, found := os.LookupEnv("WATCH_NAMESPACE")
+	if found {
+		r.OperatorNamespace = ns
+	}
+
+	// get authorino instance
+	authorinoInstance, err := r.authorinoInstance(req.NamespacedName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if authorinoInstance == nil {
+		log.Info("Authorino CR not found. returning the reconciler")
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("Found an instance of authorino", "authorinoInstanceName", authorinoInstance.Name)
+	authorinoInstance.Status.Ready = false
+
+	err = r.authorinoServices(authorinoInstance)
+	if err != nil {
+		authorinoInstance.Status.Ready = false
+		log.Error(err, "Failed to create authorino services")
+		return ctrl.Result{}, err
+	}
+
+	err = r.authorinoPermission(authorinoInstance, req.NamespacedName.Namespace)
+	if err != nil {
+		authorinoInstance.Status.Ready = false
+		log.Error(err, "Failed to create authorino permission")
+		return ctrl.Result{}, err
+	}
+
+	existingDeployment := &appsv1.Deployment{}
+	err = r.Get(context.TODO(),
+		types.NamespacedName{
+			Name:      authorinoInstance.GetName(),
+			Namespace: authorinoInstance.GetNamespace(),
+		}, existingDeployment)
+	if err != nil && errors.IsNotFound(err) {
```

Comment:
I think this could be moved to function as well

```go
func (r *AuthorinoReconciler) getAuthorinoDeployment(authorino *authorinooperatorv1beta1.Authorino) (*appsv1.Deployment, error)
  deployment := &appsv1.Deployment{}
  err = r.Get(context.TODO(), types.NamespacedName{
    Name:      authorino.GetName(),
    Namespace: authorino.GetNamespace(),
  }, deployment)
  return deployment, err
```

And then the `if` statement

```go
if existingDeployment, err := getAuthorinoDeployment(authorinoInstance); err != nil && errors.IsNotFound(err)  {
  // create the deployment
} else if existingDeployment != nil {
  // update the deployment
} else {
  // error
}
```